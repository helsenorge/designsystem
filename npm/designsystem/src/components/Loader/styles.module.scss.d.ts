export type Styles = {
  center: string;
  loader: string;
  loader__dot: string;
 ' loader__hidden-text': string;
  'loader__dot--banana': string;
  'loader__dot--black': string;
  'loader__dot--cherry': string;
  'loader__dot--kiwi': string;
  'loader__dot--large': string;
  'loader__dot--medium': string;
  'loader__dot--neutral': string;
  'loader__dot--plum': string;
  'loader__dot--small': string;
  'loader__dot--white': string;
  'loader--large': string;
  'loader--medium': string;
  'loader--small': string;
  overlay: string;
  scale: string;
  'scale-regular': string;
  'scale-reverse': string;
  translate: string;
  'translate-large': string;
  'translate-medium': string;
  'translate-small': string;
  'translate-tiny': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
