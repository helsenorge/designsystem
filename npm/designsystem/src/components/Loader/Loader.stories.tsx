import React from 'react';

import { Title as DocsTitle, Subtitle, Description, Primary, ArgsTable, PRIMARY_STORY } from '@storybook/addon-docs';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import Loader, { Overlay } from './Loader';
import { allPaletteNames } from '../../../.storybook/knobs';
import Button from '../Button/Button';
import GridExample from '../GridExample';
import Title from '../Title/Title';

export default {
  title: '@helsenorge∕designsystem-react/Components/Loader',
  component: Loader,
  parameters: {
    docs: {
      description: {
        component:
          'Loader benyttes for å vise innbygger at det lastes ned innhold, slik at innbygger må vente til innholdet er lastet ferdig.',
      },
      page: () => (
        <>
          <DocsTitle />
          <Subtitle />
          <Description />
          <Primary />
          <ArgsTable story={PRIMARY_STORY} />
        </>
      ),
    },
  },
  argTypes: {
    color: {
      control: 'select',
      options: allPaletteNames,
      defaultValue: 'blueberry',
    },
  },
} as ComponentMeta<typeof Loader>;

export const Default: ComponentStory<typeof Loader> = (args: any) => (
  <GridExample>
    <Loader {...args} />
  </GridExample>
);

export const AllSizes: ComponentStory<typeof Loader> = (args: any) => (
  <GridExample>
    <div style={{ display: 'grid', gridGap: '2rem' }}>
      <Loader {...args} size={'tiny'} />
      <Loader {...args} size={'small'} />
      <Loader {...args} size={'medium'} />
      <Loader {...args} size={'large'} />
    </div>
  </GridExample>
);

export const LoaderIsCentered: ComponentStory<typeof Loader> = (args: any) => (
  <GridExample>
    <div style={{ background: 'white' }}>
      <Loader {...args} center />
    </div>
  </GridExample>
);

export const Inline: ComponentStory<typeof Loader> = (args: any) => (
  <GridExample>
    <Title htmlMarkup="span" appearance="title1">
      {'Søker'}
    </Title>
    <Loader {...args} color="black" inline size="tiny" />
  </GridExample>
);

export const OverlayScreen: ComponentStory<typeof Loader> = (args: any) => (
  <div style={{ width: '100%', height: '200vh' }}>
    <h2>{'Fastlegen din er Arnfinn Nesset ved Orkdal sykehjem'}</h2>
    <Loader {...args} overlay={Overlay.screen} />
  </div>
);

export const OverlayParent: ComponentStory<typeof Loader> = (args: any) => (
  <div style={{ width: '100%', height: '100vh' }}>
    <h2>{'Fastlegen din er Arnfinn Nesset ved Orkdal sykehjem'}</h2>
    <div>
      <Title>{'Søknad'}</Title>
      <p style={{ color: 'red' }}>{'Søknad om greier'}</p>
      <Button>{'Lagre'}</Button>
      <Loader {...args} overlay={Overlay.parent} />
    </div>
  </div>
);
