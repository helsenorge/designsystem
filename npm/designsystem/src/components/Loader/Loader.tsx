import React from 'react';
import classNames from 'classnames';
import { v4 as uuid } from 'uuid';

import { PaletteNames } from '../../theme/palette';
import loaderStyles from './styles.module.scss';

export type LoaderColors = PaletteNames;
export type LoaderSizes = 'tiny' | 'small' | 'medium' | 'large';

interface LoaderProps {
  /** Sets the color of the loader */
  color?: LoaderColors;
  /**	Adds custom classes to the element. */
  className?: string;
  /** Changes the size of the loader. */
  size?: LoaderSizes;
  /** Sets the data-testid attribute. */
  testId?: string;
  /** Centers the loader in a container */
  center?: boolean;
  /** Loader is displayed with grey background covering the entire screen */
  overlay?: boolean;
  /**  individual id for loading icon (aria-labelledby).  */
  ariaLabelledById?: string;
  /**  individual id for loading icon (aria-labelledby).  */
  ariaLabel?: string;
}

const Loader = React.forwardRef(function LoaderForwardedRef(props: LoaderProps, ref: React.ForwardedRef<HTMLElement>) {
  const {
    color = 'neutral',
    size = 'small',
    className = '',
    testId = '',
    center,
    overlay,
    ariaLabelledById,
    ariaLabel = 'Laster inn',
  } = props;
  const isSmall = size === 'small';
  const isMedium = size === 'medium';
  const isLarge = size === 'large';

  const loaderWrapperClasses = classNames(loaderStyles['loader-wrapper'], {
    [loaderStyles['loader-wrapper--center']]: center,
    [loaderStyles['loader-wrapper--overlay']]: overlay,
  });
  const loaderClasses = classNames(
    loaderStyles.loader,
    {
      [loaderStyles['loader--small']]: isSmall,
      [loaderStyles['loader--medium']]: isMedium,
      [loaderStyles['loader--large']]: isLarge,
    },
    className
  );
  const loaderDotClasses = classNames(loaderStyles.loader__dot, {
    [loaderStyles['loader__dot--small']]: isSmall,
    [loaderStyles['loader__dot--medium']]: isMedium,
    [loaderStyles['loader__dot--large']]: isLarge,
    [loaderStyles['loader__dot--banana']]: color === 'banana',
    [loaderStyles['loader__dot--cherry']]: color === 'cherry',
    [loaderStyles['loader__dot--kiwi']]: color === 'kiwi',
    [loaderStyles['loader__dot--neutral']]: color === 'neutral',
    [loaderStyles['loader__dot--plum']]: color === 'plum',
    [loaderStyles['loader__dot--black']]: color === 'black',
    [loaderStyles['loader__dot--white']]: color === 'white',
  });

  const uniqueId = `loader${uuid()}`;

  return (
    <div role="progressbar" className={loaderWrapperClasses}>
      <div data-testid={testId} aria-labelledby={ariaLabelledById || uniqueId} className={loaderClasses}>
        <div className={loaderDotClasses} />
        <div className={loaderDotClasses} />
        <div className={loaderDotClasses} />
        <div className={loaderDotClasses} />
        {!ariaLabelledById && (
          <span id={uniqueId} className={loaderStyles['loader__hidden-text']}>
            {ariaLabel}
          </span>
        )}
      </div>
    </div>
  );
});

export default Loader;
