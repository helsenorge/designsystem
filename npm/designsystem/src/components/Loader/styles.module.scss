@import '../../scss/_palette.scss';
@import 'bootstrap/scss/mixins/_screen-reader.scss';

@mixin keyframes($name) {
  @keyframes #{$name} {
    @content;
  }
}

@mixin animate($name, $animation, $duration, $times) {
  animation: $animation $duration $times;
  animation-name: $name;
}

@mixin loaderDotScale($name, $scale-start, $scale-end) {
  @include keyframes($name) {
    0% {
      transform: scale($scale-start);
    }
    100% {
      transform: scale($scale-end);
    }
  }
  @include animate($name, scale, 0.6s, infinite);
}

@mixin loaderDotTranslate($name, $distance) {
  @include keyframes($name) {
    0% {
      transform: translate(0, 0);
    }
    100% {
      transform: translate($distance, 0);
    }
  }
  @include animate($name, translate, 0.6s, infinite);
}

.loader-wrapper {
  &--center {
    text-align: center;
  }

  &--overlay {
    z-index: 9999;
    background: $transparentGrey;
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    display: flex;
    align-items: center;
    justify-content: center;
  }
}

.loader {
  display: inline-block;
  position: relative;
  width: 8 * 5px;
  height: 8px;

  &--small {
    width: 12 * 5px;
    height: 12px;
  }

  &--medium {
    width: 18 * 5px;
    height: 18px;
  }

  &--large {
    width: 24 * 5px;
    height: 24px;
  }

  &__hidden-text {
    @include sr-only();
  }

  &__dot {
    position: absolute;
    width: 8px;
    height: 8px;
    background-color: $blueberry500;
    border-radius: 50%;

    &--small {
      width: 12px;
      height: 12px;
    }
    &--medium {
      width: 18px;
      height: 18px;
    }
    &--large {
      width: 24px;
      height: 24px;
    }

    &--banana {
      background-color: $banana500;
    }
    &--cherry {
      background-color: $cherry500;
    }
    &--kiwi {
      background-color: $kiwi500;
    }
    &--neutral {
      background-color: $neutral500;
    }
    &--plum {
      background-color: $plum500;
    }
    &--black {
      background-color: $black;
    }
    &--white {
      background-color: $white;
    }

    // Sets a form of easing animation curve
    animation-timing-function: cubic-bezier(0, 1, 1, 0);
    // Scales up the first dot
    &:nth-child(1) {
      @include loaderDotScale(scale-regular, 0, 1);
    }

    // Translates the dots rightwards depending on size
    &:nth-child(2),
    &:nth-child(3) {
      @include loaderDotTranslate(translate-tiny, 16px);
    }
    &:nth-child(3) {
      left: 16px;
    }
    &--small:nth-child(2),
    &--small:nth-child(3) {
      @include loaderDotTranslate(translate-small, 24px);
    }
    &--medium:nth-child(2),
    &--medium:nth-child(3) {
      @include loaderDotTranslate(translate-medium, 36px);
    }
    &--large:nth-child(2),
    &--large:nth-child(3) {
      @include loaderDotTranslate(translate-large, 48px);
    }
    &--small:nth-child(3) {
      left: 24px;
    }
    &--medium:nth-child(3) {
      left: 36px;
    }
    &--large:nth-child(3) {
      left: 48px;
    }

    // Scales down the last dot
    &:nth-child(4) {
      left: 32px;
      @include loaderDotScale(scale-reverse, 1, 0);
    }
    &--small:nth-child(4) {
      left: 48px;
    }
    &--medium:nth-child(4) {
      left: 72px;
    }
    &--large:nth-child(4) {
      left: 96px;
    }
  }
}
