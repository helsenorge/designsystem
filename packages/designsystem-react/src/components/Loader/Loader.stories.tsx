import React from 'react';
import {withKnobs, select} from '@storybook/addon-knobs';
import {storiesOf} from '@storybook/react';
import Loader from './Loader';
import {allPaletteNames} from '../../../.storybook/knobs';
import {withA11y} from '@storybook/addon-a11y';

const stories = storiesOf('Loader', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withA11y);

stories.add('Default', () => (
  <Loader
    size={select('Size (does not work properly)', ['tiny', 'small', 'medium', 'large'], 'tiny')}
    color={select('Color', allPaletteNames, 'blueberry')}
  />
));

stories.add('All sizes', () => (
  <div style={{display: 'grid', gridGap: '2rem'}}>
    <Loader size="tiny" color={select('Color', allPaletteNames, 'blueberry')} />
    <Loader size="small" color={select('Color', allPaletteNames, 'blueberry')} />
    <Loader size="medium" color={select('Color', allPaletteNames, 'blueberry')} />
    <Loader size="large" color={select('Color', allPaletteNames, 'blueberry')} />
  </div>
));
