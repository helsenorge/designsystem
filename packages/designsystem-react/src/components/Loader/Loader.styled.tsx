import styled, {keyframes} from 'styled-components';
import {LoaderColors, LoaderSizes} from './Loader';
import {getColor} from '../../theme/currys/color';

const loaderDot1 = keyframes`
    0% {
    transform: scale(0);
  }
  100% {
    transform: scale(1);
  }
`;

const loaderDot2 = (distance: string) => keyframes`
  0% {
    transform: translate(0, 0);
  }
  100% {
    transform: translate(${distance}, 0);
  }
`;

const loaderDot3 = keyframes`
  0% {
    transform: scale(1);
  }
  100% {
    transform: scale(0);
  }
`;

const dotSizes = {
  tiny: 8,
  small: 12,
  medium: 18,
  large: 24,
};

export const StyledLoaderDot = styled('div')<{color: LoaderColors; size: LoaderSizes}>`
  position: absolute;
  width: ${props => `${dotSizes[props.size]}px`};
  height: ${props => `${dotSizes[props.size]}px`};
  border-radius: 50%;
  background-color: ${props => getColor(props.color, 500)};
  animation-timing-function: cubic-bezier(0, 1, 1, 0);
  &:nth-child(1) {
    animation: ${loaderDot1} 0.6s infinite;
  }
  &:nth-child(2) {
    animation: ${props => loaderDot2(`${dotSizes[props.size] * 2}px`)} 0.6s infinite;
  }
  &:nth-child(3) {
    left: ${props => `${dotSizes[props.size] * 2}px`};
    animation: ${props => loaderDot2(`${dotSizes[props.size] * 2}px`)} 0.6s infinite;
  }
  &:nth-child(4) {
    left: ${props => `${dotSizes[props.size] * 4}px`};
    animation: ${loaderDot3} 0.6s infinite;
  }
`;

export const StyledLoader = styled('div')<{size: LoaderSizes}>`
  display: inline-block;
  position: relative;
  width: ${props => `${dotSizes[props.size] * 5}px`};
  height: ${props => `${dotSizes[props.size]}px`};
`;
