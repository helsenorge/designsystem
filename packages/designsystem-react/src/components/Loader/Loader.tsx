import React from 'react';
import {StyledLoader, StyledLoaderDot} from './Loader.styled';
import {PaletteNames} from '../../theme/palette';

export type LoaderColors = PaletteNames;
export type LoaderSizes = 'tiny' | 'small' | 'medium' | 'large';

interface LoaderProps {
  color?: LoaderColors;
  className?: string;
  size?: LoaderSizes;
  testId?: string;
}

const Loader = React.forwardRef((props: LoaderProps, ref: any) => {
  const {color = 'neutral', size = 'small', className = '', testId = ''} = props;
  return (
    <StyledLoader data-testid={testId} className={className} size={size}>
      <StyledLoaderDot color={color} size={size} />
      <StyledLoaderDot color={color} size={size} />
      <StyledLoaderDot color={color} size={size} />
      <StyledLoaderDot color={color} size={size} />
    </StyledLoader>
  );
});

export default Loader;
