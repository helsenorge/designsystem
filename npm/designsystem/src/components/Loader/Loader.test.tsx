import React from 'react';
import { render, screen } from '@testing-library/react';
import Loader from './Loader';

jest.mock('uuid', () => {
  const uuidGen = (): string => `-unik-id`;
  return { v4: uuidGen };
});

describe('Gitt at en loader skal vises ', (): void => {
  describe('Når at en loader rendres ', (): void => {
    test('Så vises loaderen', (): void => {
      const { container } = render(<Loader />);
      expect(container).toMatchSnapshot();
    });

    test('Så er alle fire loaderDots riktig plassert', (): void => {
      render(<Loader testId={'loaderDotTest'} />);
      const loaderDotsWrapper = screen.getByTestId('loaderDotTest');

      expect(loaderDotsWrapper.children[0].className).toBe('loader__dot loader__dot--small loader__dot--neutral');
      expect(loaderDotsWrapper.children[1].className).toBe('loader__dot loader__dot--small loader__dot--neutral');
      expect(loaderDotsWrapper.children[2].className).toBe('loader__dot loader__dot--small loader__dot--neutral');
      expect(loaderDotsWrapper.children[3].className).toBe('loader__dot loader__dot--small loader__dot--neutral');
    });

    test('Så har loader progressbar role', (): void => {
      render(<Loader overlay />);
      expect(screen.getByRole('progressbar')).toBeVisible();
    });
  });

  describe('Når center prop er true ', (): void => {
    test('Så sentreres loader', (): void => {
      const { container } = render(<Loader center />);
      expect(container.firstChild).toHaveClass('loader-wrapper--center');
    });
  });

  describe('Når overlay prop er true', (): void => {
    test('Så har loaderen en overlay', (): void => {
      const { container } = render(<Loader overlay />);
      expect(container.firstChild).toHaveClass('loader-wrapper--overlay');
      expect(container.firstChild.firstChild).toHaveAttribute('aria-labelledby', 'loader-unik-id');
    });
  });

  describe('Når ariaLabelledById er satt', (): void => {
    test('Så settes attributten aria-labelledby til det samme', (): void => {
      const { container } = render(<Loader ariaLabelledById="aria-test" overlay />);
      expect(container.firstChild).toHaveClass('loader-wrapper--overlay');
      expect(container.firstChild.firstChild).toHaveAttribute('aria-labelledby', 'aria-test');
    });
  });
});
