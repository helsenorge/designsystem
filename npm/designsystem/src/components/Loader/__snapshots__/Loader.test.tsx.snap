// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt at en loader skal vises  Når at en loader rendres  Så vises loaderen 1`] = `
<div>
  <div
    class="loader-wrapper"
    role="progressbar"
  >
    <div
      aria-labelledby="loader-unik-id"
      class="loader loader--small"
      data-testid=""
    >
      <div
        class="loader__dot loader__dot--small loader__dot--neutral"
      />
      <div
        class="loader__dot loader__dot--small loader__dot--neutral"
      />
      <div
        class="loader__dot loader__dot--small loader__dot--neutral"
      />
      <div
        class="loader__dot loader__dot--small loader__dot--neutral"
      />
      <span
        class="loader__hidden-text"
        id="loader-unik-id"
      >
        Laster inn
      </span>
    </div>
  </div>
</div>
`;
